src/midje/parsing/0_to_fact_form/generative.clj

```diff
@@ -71,19 +73,86 @@
              [run# passes#] (run-for-all (list ~num-tests
                                                prop#
                                                ~@quick-check-opts))]
+         (if (:result run#)
+           (dotimes [_# (/ passes# ~num-tests)]
+             (emission/pass))
+           (run-with-smallest ~fact-fn-name '~prop-names run#))
+         (boolean (:result run#))))))
+
+(declare roll-up-bindings)
+
+(defn bind-expr [gen-name gen-expr inner-body other-bindings]
+  (roll-up-bindings
+    other-bindings
+    (list 'clojure.test.check.generators/bind
+          gen-expr
+          (list `fn [gen-name] inner-body))))
+
+(defn let-expr [fmap-bindings inner-body other-bindings]
+  (roll-up-bindings
+    other-bindings
+    `(let ~fmap-bindings
+       ~inner-body)))
+
+(defn roll-up-bindings [bindings inner-body]
+  (let [[fst-binding & rst-bindings] bindings
+        [lhs rhs] fst-binding]
+    (cond
+       (empty? bindings) inner-body
+       (= :let lhs)  (let-expr rhs inner-body rst-bindings)
+       :else (bind-expr lhs rhs inner-body rst-bindings))))
+
+
+(defn roll-up [bindings gen-names check-fn-name]
+  (roll-up-bindings (reverse bindings)
+                    `(gen/return {:function ~check-fn-name
+                                  :result   (~check-fn-name ~@gen-names)
+                                  :args     (list ~@gen-names)})))
+
+(defn variables [bindings]
+  (->> bindings
+       (mapcat (fn [[lhs rhs]]
+                 (cond (= :let lhs)  (->> rhs (partition 2) (map first))
+                       :else [lhs])))
+       vec))
+
+(defn- build-gen-let-parser [form]
+  (fn []
+    (let [[metadata forms] (parse-metadata/separate-metadata form)
+          [prop-names prop-values
+           opts checks] (parse-for-all-form form forms)
+          num-tests        (or (:num-tests opts) 10)
+          quick-check-opts (->> (select-keys opts [:seed :max-size])
+                                (into [])
+                                flatten)
+          fact-fn-name     (gensym "fact-fn")
+          bindings         (map vector prop-names prop-values)
+          vars             (variables bindings)
+          prop             (roll-up bindings vars fact-fn-name)]
+      `(let [~fact-fn-name  (fn ~vars
+                              ~(parse-facts/wrap-fact-around-body
+                                 metadata checks))
+             [run# passes#] (run-for-all (list ~num-tests
+                                               ~prop
+                                               ~@quick-check-opts))]
          (if (:pass? run#)
            (dotimes [_# (/ passes# ~num-tests)]
              (emission/pass))
-           (run-with-smallest fact-fn# '~prop-names run#))
-         (:pass? run#)))))
+           (run-with-smallest ~fact-fn-name '~vars run#))
+         (boolean (:pass? run#))))))
 
 (defn parse-for-all [form]
   (error/parse-and-catch-failure form (build-parser form)))
```

Comment:
might be cool to rename `build-parser` to `build-for-all-parser` (I can do it later too)